### PR TITLE
TT-7351 changed precedence on how plugins are picked

### PR DESCRIPTION
**Pull request by sredxny**
## Description

Changed precedence on how goplugins are picked to be loaded. It used to be: 

1. exact name provided by user
2. new naming convention with unprefixed gw version
3. new naming convention

Now it was modified to be:

1. New naming convention
2. New naminf convention with unprefixed gw version
3. exact name provided by user

In this way it will always attempt to load a compatible plugin and will avoid crashes when is found a file that is not intended to be used by that version of gw

## Related Issue

https://tyktech.atlassian.net/browse/TT-7351

## How This Has Been Tested

- Added unit tests
- Manual testing. In a single directory have 3 plugins files which name is: `myplugin.so`, `myplugin_v4.3.0_darwin_amd64.so`,`myplugin_4.3.0_darwin_amd64.so`
- Create an api that uses a goplugin which name is `myplugin.so`
- The plugin `myplugin_v4.3.0_darwin_amd64.so` should be picked
- Remove the file `myplugin_v4.3.0_darwin_amd64.so` (if you use bundles then remove any local cache, check the directory set in the gw config `middleware_path`+bundles)
- Restart the gateway , the file `myplugin_4.3.0_darwin_amd64.so` should be loaded
- Remove the file `myplugin_4.3.0_darwin_amd64.so` and restart the gateway
- The file `myplugin.so` should be loaded

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Refactoring or add test (improvements in base code or adds test coverage to functionality)

## Checklist

- [ ] I ensured that the documentation is up to date
- [ ] I explained why this PR updates go.mod in detail with reasoning why it's required
- [ ] I would like a code coverage CI quality gate exception and have explained why

### CODE REVIEW

**Comment by Tyk-ITS** [comment]
**API tests result: success** :white_check_mark: 
Branch used: refs/pull/4576/merge
Commit:  
Triggered by: pull_request (@sredxny)
[Execution page](https://github.com/TykTechnologies/tyk/actions/runs/3760533510)

File: goplugin/plugin_name_builder.go [hunk 1]
```diff
@@ -12,19 +12,25 @@ var pluginStorage storage
...
+	if !strings.HasPrefix(version, "v") {
+		newNamingFormat = getPluginNameFromTykVersion(version, path)
+
+		if pluginStorage.fileExist(newNamingFormat) {
+			return newNamingFormat
 		}
-		path = newPath
 	}
 
+	// 3. attempt to load the exact name provided
```
Comment (tbuchaillot) [inline goplugin/plugin_name_builder.go:33]: @sredxny can we check if `path` file exists at this point?

File: goplugin/plugin_name_builder.go [hunk 2]
```diff
@@ -12,19 +12,25 @@ var pluginStorage storage
 // GetPluginFileNameToLoad check which file to load based on name, tyk version, os and architecture
 // but it also takes care of returning the name of the file that exists
 func GetPluginFileNameToLoad(path string, version string) string {
```
Comment (tbuchaillot) [inline goplugin/plugin_name_builder.go:14]: @sredxny can we make this func returns an error if the file file doesn't exist in any of the paths?

**Review by tbuchaillot — CHANGES_REQUESTED** [review]

**Comment by Tyk-ITS** [comment]
**API tests result: success** :white_check_mark: 
Branch used: refs/pull/4576/merge
Commit: fb2538dd738a308a350ab59b2be0b93e0cf7498b 
Triggered by: pull_request (@tbuchaillot)
[Execution page](https://github.com/TykTechnologies/tyk/actions/runs/3764977047)

File: goplugin/plugin_name_builder.go (hunk 1 above)
Comment (sredxny) [reply goplugin/plugin_name_builder.go:34]: done

File: goplugin/plugin_name_builder.go (hunk 2 above)
Comment (sredxny) [reply goplugin/plugin_name_builder.go]: done

**Review by tbuchaillot — APPROVED** [review]
LGTM

**Comment by Tyk-ITS** [comment]
**API tests result: failure** :no_entry_sign: 
Branch used: refs/pull/4576/merge
Commit: 85276188ed2c2d20cbaa9538998dd5f667f98250 
Triggered by: pull_request (@sredxny)
[Execution page](https://github.com/TykTechnologies/tyk/actions/runs/3765532111)

**Comment by Tyk-ITS** [comment]
**API tests result: failure** :no_entry_sign: 
Branch used: refs/pull/4576/merge
Commit: 28a0ef171b662b5e0e175893917374f890ec0716 
Triggered by: pull_request (@sredxny)
[Execution page](https://github.com/TykTechnologies/tyk/actions/runs/3765550711)

**Comment by Tyk-ITS** [comment]
**API tests result: success** :white_check_mark: 
Branch used: refs/pull/4576/merge
Commit: 3da1b32477b2ce4d244f4de9cd04b36b3af3ac96 
Triggered by: pull_request (@sredxny)
[Execution page](https://github.com/TykTechnologies/tyk/actions/runs/3765884027)

**Comment by Tyk-ITS** [comment]
**API tests result: success** :white_check_mark: 
Branch used: refs/pull/4576/merge
Commit: ee8d6bf4814bb0b799d51f66ef20fe3eb191a144 
Triggered by: pull_request (@sredxny)
[Execution page](https://github.com/TykTechnologies/tyk/actions/runs/3765919612)

**Comment by Tyk-ITS** [comment]
**API tests result: failure** :no_entry_sign: 
Branch used: refs/pull/4576/merge
Commit: 9b09e3d50e7f203806f6dab29cc98d82899852fc 
Triggered by: pull_request (@sredxny)
[Execution page](https://github.com/TykTechnologies/tyk/actions/runs/3765953865)

**Review by mativm02 — APPROVED** [review]
lgtm!

**Comment by Tyk-ITS** [comment]
**API tests result: failure** :no_entry_sign: 
Branch used: refs/heads/master
Commit: 8e0a9e72748513ee91f69a85ddd887a273d2a868 TT-7351 changed precedence on how plugins are picked (#4576)



## Description

Changed precedence on how goplugins are picked to be loaded. It used to
be:

1. exact name provided by user
2. new naming convention with unprefixed gw version
3. new naming convention

Now it was modified to be:

1. New naming convention
2. New naminf convention with unprefixed gw version
3. exact name provided by user

In this way it will always attempt to load a compatible plugin and will
avoid crashes when is found a file that is not intended to be used by
that version of gw

## Related Issue

https://tyktech.atlassian.net/browse/TT-7351

## Motivation and Context



## How This Has Been Tested

- Added unit tests
- Manual testing. In a single directory have 3 plugins files which name
is: `myplugin.so`,
`myplugin_v4.3.0_darwin_amd64.so`,`myplugin_4.3.0_darwin_amd64.so`
- Create an api that uses a goplugin which name is `myplugin.so`
- The plugin `myplugin_v4.3.0_darwin_amd64.so` should be picked
- Remove the file `myplugin_v4.3.0_darwin_amd64.so` (if you use bundles
then remove any local cache, check the directory set in the gw config
`middleware_path`+bundles)
- Restart the gateway , the file `myplugin_4.3.0_darwin_amd64.so` should
be loaded
- Remove the file `myplugin_4.3.0_darwin_amd64.so` and restart the
gateway
- The file `myplugin.so` should be loaded

## Screenshots (if appropriate)

## Types of changes



- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing
functionality to change)
- [ ] Refactoring or add test (improvements in base code or adds test
coverage to functionality)

## Checklist





- [ ] I ensured that the documentation is up to date
- [ ] I explained why this PR updates go.mod in detail with reasoning
why it's required
- [ ] I would like a code coverage CI quality gate exception and have
explained why

Co-authored-by: Tomas Buchaillot <tombuchaillot89@gmail.com>
Triggered by: push (@sredxny)
[Execution page](https://github.com/TykTechnologies/tyk/actions/runs/3766130114)

**Pull request merged**